<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 4.2.1">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/net2.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/net1.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">
  <meta name="google-site-verification" content="LASEpzFfcnNFnu_8Sy9VvRqObpPtUGArZOdqPLDPTp8">

<link rel="stylesheet" href="/css/main.css">

<link rel="stylesheet" href="//fonts.googleapis.com/css?family=sans-serif:300,300italic,400,400italic,700,700italic|consolas:300,300italic,400,400italic,700,700italic&display=swap&subset=latin,latin-ext">
<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">
  <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox@3/dist/jquery.fancybox.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"geng-lee.github.io","root":"/","scheme":"Muse","version":"8.0.0-rc.3","exturl":true,"sidebar":{"position":"right","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":true,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="Introduction   10X Adult Mouse Brain In this example, we will be analyzing a dataset of 5K cells from the adult mouse brain available from 10X genomics.">
<meta property="og:type" content="article">
<meta property="og:title" content="Analysis of single-cell ATAC-seq data using SnapATAC software (2)Ôºö10x Adult Mouse Brain">
<meta property="og:url" content="https://geng-lee.github.io/posts/34331839/index.html">
<meta property="og:site_name" content="GengLee&#39;s Blog">
<meta property="og:description" content="Introduction   10X Adult Mouse Brain In this example, we will be analyzing a dataset of 5K cells from the adult mouse brain available from 10X genomics.">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-1-QualityControl.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-2-BinCovDist.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-3-eigs_scatter_plot.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-4-Visulization.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-5-gene_plot.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-6-Visulization_tsne.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-7-cluster_tree.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-8-DARs.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-9-DARs_all.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-10-homer_motif.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-11-chromVAR_2B.jpg">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-12-chromVAR_2C.jpg">
<meta property="article:published_time" content="2020-06-25T01:05:25.000Z">
<meta property="article:modified_time" content="2020-08-02T15:29:07.000Z">
<meta property="article:author" content="Geng Lee">
<meta property="article:tag" content="R">
<meta property="article:tag" content="scRNA">
<meta property="article:tag" content="ATAC">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-1-QualityControl.jpg">

<link rel="canonical" href="https://geng-lee.github.io/posts/34331839/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>Analysis of single-cell ATAC-seq data using SnapATAC software (2)Ôºö10x Adult Mouse Brain | GengLee's Blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<link rel="alternate" href="/atom.xml" title="GengLee's Blog" type="application/atom+xml">
</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
        <span class="toggle-line toggle-line-first"></span>
        <span class="toggle-line toggle-line-middle"></span>
        <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">GengLee's Blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">üòè ÂÅöÂÄãËø∑‰∫∫ÁöÑÊ∑∑Ëõã Ê°ÄÈ©ÅËÄå‰∏çÂ§±ÈãíÁØÑ</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>



<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>Categories</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

  <span class="exturl github-corner" data-url="aHR0cHM6Ly9naXRodWIuY29tL2dlbmctbGVl" title="Follow me on GitHub" aria-label="Follow me on GitHub"><svg width="80" height="80" viewBox="0 0 250 250" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></span>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://geng-lee.github.io/posts/34331839/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.jpg">
      <meta itemprop="name" content="Geng Lee">
      <meta itemprop="description" content="üòè To be a better me">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="GengLee's Blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          Analysis of single-cell ATAC-seq data using SnapATAC software (2)Ôºö10x Adult Mouse Brain
        </h1>

        <div class="post-meta">

          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2020-06-25 09:05:25" itemprop="dateCreated datePublished" datetime="2020-06-25T09:05:25+08:00">2020-06-25</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2020-08-02 23:29:07" itemprop="dateModified" datetime="2020-08-02T23:29:07+08:00">2020-08-02</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/scRNA/" itemprop="url" rel="index"><span itemprop="name">scRNA</span></a>
                </span>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>18k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>16 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h1 id="introduction"><a class="markdownIt-Anchor" href="#introduction"></a> Introduction</h1>
<hr />
<div class="note info no-icon"><h2 id="10x-adult-mouse-brain"><a class="markdownIt-Anchor" href="#10x-adult-mouse-brain"></a> 10X Adult Mouse Brain</h2>
<p>In this example, we will be analyzing a dataset of 5K cells from the adult mouse brain available from 10X genomics.</p>
</div>
<br>
<a id="more"></a>
<h1 id="analysis-content"><a class="markdownIt-Anchor" href="#analysis-content"></a> Analysis content</h1>
<hr />
<h2 id="table-of-contents"><a class="markdownIt-Anchor" href="#table-of-contents"></a> Table of Contents</h2>
<ul>
<li><a href="#introduction"><font color="#1269d3">Introduction</font></a>
<ul>
<li><a href="#10x-adult-mouse-brain"><font color="#1269d3">10X Adult Mouse Brain</font></a></li>
</ul>
</li>
<li><a href="#analysis-content"><font color="#1269d3">Analysis content</font></a>
<ul>
<li>
<p><a href="#table-of-contents"><font color="#1269d3">Table of Contents</font></a></p>
</li>
<li>
<p><a href="#step-1-barcode-selection"><font color="#1269d3">Step 1. Barcode selection</font></a></p>
</li>
<li>
<p><a href="#step-2-add-cell-by-bin-matrix"><font color="#1269d3">Step 2. Add cell-by-bin matrix</font></a></p>
</li>
<li>
<p><a href="#step-3-matrix-binarization"><font color="#1269d3">Step 3. Matrix binarization</font></a></p>
</li>
<li>
<p><a href="#step-4-bin-filtering"><font color="#1269d3">Step 4. Bin filtering</font></a></p>
</li>
<li>
<p><a href="#step-5-dimensionality-reduction"><font color="#1269d3">Step 5. Dimensionality reduction</font></a></p>
</li>
<li>
<p><a href="#step-6-determine-significant-components"><font color="#1269d3">Step 6. Determine significant components</font></a></p>
</li>
<li>
<p><a href="#step-7-graph-based-clustering"><font color="#1269d3">Step 7. Graph-based clustering</font></a></p>
</li>
<li>
<p><a href="#step-8-visualization"><font color="#1269d3">Step 8. Visualization</font></a></p>
</li>
<li>
<p><a href="#step-9-gene-based-annotation"><font color="#1269d3">Step 9. Gene based annotation</font></a></p>
</li>
<li>
<p><a href="#step-10-heretical-clustering"><font color="#1269d3">Step 10. Heretical clustering</font></a></p>
</li>
<li>
<p><a href="#step-11-identify-peaks"><font color="#1269d3">Step 11. Identify peaks</font></a></p>
</li>
<li>
<p><a href="#step-12-create-a-cell-by-peak-matrix"><font color="#1269d3">Step 12. Create a cell-by-peak matrix</font></a></p>
</li>
<li>
<p><a href="#step-13-add-cell-by-peak-matrix"><font color="#1269d3">Step 13. Add cell-by-peak matrix</font></a></p>
</li>
<li>
<p><a href="#step-14-identify-differentially-accessible-regions"><font color="#1269d3">Step 14. Identify differentially accessible regions</font></a></p>
</li>
<li>
<p><a href="#step-15-motif-analysis-identifies-master-regulators"><font color="#1269d3">Step 15. Motif analysis identifies master regulators</font></a></p>
</li>
<li>
<p><a href="#step-16-great-analysis"><font color="#1269d3">Step 16. GREAT analysis</font></a></p>
</li>
</ul>
</li>
</ul>
<h2 id="a-namebarcode_selectionastep-1-barcode-selection"><a class="markdownIt-Anchor" href="#a-namebarcode_selectionastep-1-barcode-selection"></a> <a name="barcode_selection"></a><strong>Step 1. Barcode selection</strong></h2>
<p>We select high-quality barcodes based on two criteria:<br> 1) number of unique fragments;<br> 2) fragments in promoter ratio;</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; <span class="keyword">library</span>(SnapATAC);</span><br><span class="line">&gt; x.sp = createSnap(</span><br><span class="line">    file=<span class="string">"atac_v1_adult_brain_fresh_5k.snap"</span>,</span><br><span class="line">    sample=<span class="string">"atac_v1_adult_brain_fresh_5k"</span>,</span><br><span class="line">    num.cores=<span class="number">1</span></span><br><span class="line">  );</span><br><span class="line">&gt; barcodes = read.csv(</span><br><span class="line">    <span class="string">"atac_v1_adult_brain_fresh_5k_singlecell.csv"</span>,</span><br><span class="line">    head=<span class="literal">TRUE</span></span><br><span class="line">  );</span><br><span class="line">&gt; barcodes = barcodes[<span class="number">2</span>:nrow(barcodes),];</span><br><span class="line">&gt; promoter_ratio = (barcodes$promoter_region_fragments+<span class="number">1</span>) / (barcodes$passed_filters + <span class="number">1</span>);</span><br><span class="line">&gt; UMI = log(barcodes$passed_filters+<span class="number">1</span>, <span class="number">10</span>);</span><br><span class="line">&gt; data = data.frame(UMI=UMI, promoter_ratio=promoter_ratio);</span><br><span class="line">&gt; barcodes$promoter_ratio = promoter_ratio;</span><br><span class="line">&gt; <span class="keyword">library</span>(viridisLite);</span><br><span class="line">&gt; <span class="keyword">library</span>(ggplot2);</span><br><span class="line">&gt; p1 = ggplot(</span><br><span class="line">    data, </span><br><span class="line">    aes(x= UMI, y= promoter_ratio)) + </span><br><span class="line">    geom_point(size=<span class="number">0.1</span>, col=<span class="string">"grey"</span>) +</span><br><span class="line">    theme_classic() +</span><br><span class="line">    ggtitle(<span class="string">"10X Fresh Adult Brain"</span>) +</span><br><span class="line">    ylim(<span class="number">0</span>, <span class="number">1</span>) + xlim(<span class="number">0</span>, <span class="number">6</span>) +</span><br><span class="line">    labs(x = <span class="string">"log10(UMI)"</span>, y=<span class="string">"promoter ratio"</span>) </span><br><span class="line">&gt; p1 </span><br><span class="line">&gt; barcodes.sel = barcodes[which(UMI &gt;= <span class="number">3</span> &amp; UMI &lt;= <span class="number">5</span> &amp; promoter_ratio &gt;= <span class="number">0.15</span> &amp; promoter_ratio &lt;= <span class="number">0.6</span>),];</span><br><span class="line">&gt; rownames(barcodes.sel) = barcodes.sel$barcode;</span><br><span class="line">&gt; x.sp = x.sp[which(x.sp@barcode %<span class="keyword">in</span>% barcodes.sel$barcode),];</span><br><span class="line">&gt; x.sp@metaData = barcodes.sel[x.sp@barcode,];</span><br><span class="line">&gt; x.sp</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## number <span class="keyword">of</span> barcodes: <span class="number">4100</span></span><br><span class="line">## number <span class="keyword">of</span> bins: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> genes: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> peaks: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> motifs: <span class="number">0</span></span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-1-QualityControl.jpg" width="400" height="400" />  
<h2 id="a-nameadd_bmatastep-2-add-cell-by-bin-matrix"><a class="markdownIt-Anchor" href="#a-nameadd_bmatastep-2-add-cell-by-bin-matrix"></a> <a name="add_bmat"></a><strong>Step 2. Add cell-by-bin matrix</strong></h2>
<p>Next, we add the cell-by-bin matrix of 5kb resolution to the snap object. This function will automatically read the cell-by-bin matrix and add it to <mark class="label default"><font color=c52950>bmat</font></mark> slot of snap object.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment"># show what bin sizes exist in atac_v1_adult_brain_fresh_5k.snap file</span></span><br><span class="line">&gt; showBinSizes(<span class="string">"atac_v1_adult_brain_fresh_5k.snap"</span>);</span><br><span class="line">[<span class="number">1</span>] <span class="number">1000</span> <span class="number">5000</span> <span class="number">10000</span></span><br><span class="line">&gt; x.sp = addBmatToSnap(x.sp, bin.size=<span class="number">5000</span>, num.cores=<span class="number">1</span>);</span><br></pre></td></tr></table></figure>
<h2 id="a-namemake_binaryastep-3-matrix-binarization"><a class="markdownIt-Anchor" href="#a-namemake_binaryastep-3-matrix-binarization"></a> <a name="make_binary"></a><strong>Step 3. Matrix binarization</strong></h2>
<p>We will convert the cell-by-bin count matrix to a binary matrix. Some items in the count matrix have abnormally high coverage perhaps due to the alignment errors. Therefore, we next remove 0.1% items of the highest coverage in the count matrix and then convert the remaining non-zero items to 1.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; x.sp = makeBinary(x.sp, mat=<span class="string">"bmat"</span>);</span><br></pre></td></tr></table></figure>
<h2 id="a-namebin_filterastep-4-bin-filtering"><a class="markdownIt-Anchor" href="#a-namebin_filterastep-4-bin-filtering"></a> <a name="bin_filter"></a><strong>Step 4. Bin filtering</strong></h2>
<p>First, we filter out any bins overlapping with the ENCODE blacklist to prevent from potential artifacts.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; system(<span class="string">"wget http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz"</span>);</span><br><span class="line">&gt; <span class="keyword">library</span>(GenomicRanges);</span><br><span class="line">&gt; black_list = read.table(<span class="string">"mm10.blacklist.bed.gz"</span>);</span><br><span class="line">&gt; black_list.gr = GRanges(</span><br><span class="line">    black_list[,<span class="number">1</span>], </span><br><span class="line">    IRanges(black_list[,<span class="number">2</span>], black_list[,<span class="number">3</span>])</span><br><span class="line">  );</span><br><span class="line">&gt; idy = queryHits(findOverlaps(x.sp@feature, black_list.gr));</span><br><span class="line">&gt; <span class="keyword">if</span>(length(idy) &gt; <span class="number">0</span>)&#123;x.sp = x.sp[,-idy, mat=<span class="string">"bmat"</span>]&#125;;</span><br><span class="line">&gt; x.sp</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## number <span class="keyword">of</span> barcodes: <span class="number">4100</span></span><br><span class="line">## number <span class="keyword">of</span> bins: <span class="number">546103</span></span><br><span class="line">## number <span class="keyword">of</span> genes: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> peaks: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> motifs: <span class="number">0</span></span><br></pre></td></tr></table></figure>
<p>Second, we remove unwanted chromosomes.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; chr.exclude = seqlevels(x.sp@feature)[grep(<span class="string">"random|chrM"</span>, seqlevels(x.sp@feature))];</span><br><span class="line">&gt; idy = grep(paste(chr.exclude, collapse=<span class="string">"|"</span>), x.sp@feature);</span><br><span class="line">&gt; <span class="keyword">if</span>(length(idy) &gt; <span class="number">0</span>)&#123;x.sp = x.sp[,-idy, mat=<span class="string">"bmat"</span>]&#125;;</span><br><span class="line">&gt; x.sp</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## number <span class="keyword">of</span> barcodes: <span class="number">4100</span></span><br><span class="line">## number <span class="keyword">of</span> bins: <span class="number">545183</span></span><br><span class="line">## number <span class="keyword">of</span> genes: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> peaks: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> motifs: <span class="number">0</span></span><br></pre></td></tr></table></figure>
<p>Third, the bin coverage roughly obeys a log normal distribution. We remove the top 5% bins that overlap with invariant features such as promoters of the house keeping genes.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; bin.cov = log10(Matrix::colSums(x.sp@bmat)+<span class="number">1</span>);</span><br><span class="line">&gt; hist(</span><br><span class="line">    bin.cov[bin.cov &gt; <span class="number">0</span>], </span><br><span class="line">    xlab=<span class="string">"log10(bin cov)"</span>, </span><br><span class="line">    main=<span class="string">"log10(Bin Cov)"</span>, </span><br><span class="line">    col=<span class="string">"lightblue"</span>, </span><br><span class="line">    xlim=c(<span class="number">0</span>, <span class="number">5</span>)</span><br><span class="line">  );</span><br><span class="line">&gt; bin.cutoff = quantile(bin.cov[bin.cov &gt; <span class="number">0</span>], <span class="number">0.95</span>);</span><br><span class="line">&gt; idy = which(bin.cov &lt;= bin.cutoff &amp; bin.cov &gt; <span class="number">0</span>);</span><br><span class="line">&gt; x.sp = x.sp[, idy, mat=<span class="string">"bmat"</span>];</span><br><span class="line">&gt; x.sp</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## number <span class="keyword">of</span> barcodes: <span class="number">4100</span></span><br><span class="line">## number <span class="keyword">of</span> bins: <span class="number">474624</span></span><br><span class="line">## number <span class="keyword">of</span> genes: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> peaks: <span class="number">0</span></span><br><span class="line">## number <span class="keyword">of</span> motifs: <span class="number">0</span></span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-2-BinCovDist.jpg" width="400" height="400" />  
<h2 id="a-namediffusion_mapsastep-5-dimensionality-reduction"><a class="markdownIt-Anchor" href="#a-namediffusion_mapsastep-5-dimensionality-reduction"></a> <a name="diffusion_maps"></a><strong>Step 5. Dimensionality reduction</strong></h2>
<p>We compute diffusion maps for dimentionality reduction.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; x.sp = runDiffusionMaps(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    input.mat=<span class="string">"bmat"</span>, </span><br><span class="line">    num.eigs=<span class="number">50</span></span><br><span class="line">  );</span><br></pre></td></tr></table></figure>
<h2 id="a-namepc_selectastep-6-determine-significant-components"><a class="markdownIt-Anchor" href="#a-namepc_selectastep-6-determine-significant-components"></a> <a name="pc_select"></a><strong>Step 6. Determine significant components</strong></h2>
<p>We next determine the number of reduced dimensions to include for downstream analysis. We use an ad hoc method by simply looking at a pairwise plot and select the number of dimensions in which the scatter plot starts looking like a blob. In the below example, we choose the first 20 dimensions.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; plotDimReductPW(</span><br><span class="line">    obj=x.sp, </span><br><span class="line">    eigs.dims=<span class="number">1</span>:<span class="number">50</span>,</span><br><span class="line">    point.size=<span class="number">0.3</span>,</span><br><span class="line">    point.color=<span class="string">"grey"</span>,</span><br><span class="line">    point.shape=<span class="number">19</span>,</span><br><span class="line">    point.alpha=<span class="number">0.6</span>,</span><br><span class="line">    down.sample=<span class="number">5000</span>,</span><br><span class="line">    pdf.file.name=<span class="literal">NULL</span>, </span><br><span class="line">    pdf.height=<span class="number">7</span>, </span><br><span class="line">    pdf.width=<span class="number">7</span></span><br><span class="line">  );</span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-3-eigs_scatter_plot.jpg" width="900" height="900" />  
<h2 id="a-nameclusterastep-7-graph-based-clustering"><a class="markdownIt-Anchor" href="#a-nameclusterastep-7-graph-based-clustering"></a> <a name="cluster"></a><strong>Step 7. Graph-based clustering</strong></h2>
<p>Using the selected significant dimensions, we next construct a K Nearest Neighbor (KNN) Graph (k=15). Each cell is a node and the k-nearest neighbors of each cell are identified according to the Euclidian distance and edges are draw between neighbors in the graph.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; x.sp = runKNN(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    eigs.dims=<span class="number">1</span>:<span class="number">20</span>,</span><br><span class="line">    k=<span class="number">15</span></span><br><span class="line">  );</span><br><span class="line">&gt; x.sp=runCluster(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    tmp.folder=tempdir(),</span><br><span class="line">    louvain.lib=<span class="string">"R-igraph"</span>,</span><br><span class="line">    seed.use=<span class="number">10</span></span><br><span class="line">  );</span><br><span class="line">&gt; x.sp@metaData$cluster = x.sp@cluster;</span><br></pre></td></tr></table></figure>
<h2 id="a-namevizastep-8-visualization"><a class="markdownIt-Anchor" href="#a-namevizastep-8-visualization"></a> <a name="viz"></a><strong>Step 8. Visualization</strong></h2>
<p>SnapATAC visualizes and explores the data using tSNE (FI-tsne) or UMAP. In this example, we compute the t-SNE embedding. We next project the sequencing depth or other bias onto the t-SNE embedding.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; x.sp = runViz(</span><br><span class="line">    obj=x.sp, </span><br><span class="line">    tmp.folder=tempdir(),</span><br><span class="line">    dims=<span class="number">2</span>,</span><br><span class="line">    eigs.dims=<span class="number">1</span>:<span class="number">20</span>, </span><br><span class="line">    method=<span class="string">"Rtsne"</span>,</span><br><span class="line">    seed.use=<span class="number">10</span></span><br><span class="line">  );</span><br><span class="line">&gt; par(mfrow = c(<span class="number">2</span>, <span class="number">2</span>));</span><br><span class="line">&gt; plotViz(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    method=<span class="string">"tsne"</span>, </span><br><span class="line">    main=<span class="string">"10X Brain Cluster"</span>,</span><br><span class="line">    point.color=x.sp@cluster, </span><br><span class="line">    point.size=<span class="number">1</span>, </span><br><span class="line">    point.shape=<span class="number">19</span>, </span><br><span class="line">    point.alpha=<span class="number">0.8</span>, </span><br><span class="line">    text.add=<span class="literal">TRUE</span>,</span><br><span class="line">    text.size=<span class="number">1.5</span>,</span><br><span class="line">    text.color=<span class="string">"black"</span>,</span><br><span class="line">    text.halo.add=<span class="literal">TRUE</span>,</span><br><span class="line">    text.halo.color=<span class="string">"white"</span>,</span><br><span class="line">    text.halo.width=<span class="number">0.2</span>,</span><br><span class="line">    down.sample=<span class="number">10000</span>,</span><br><span class="line">    legend.add=<span class="literal">FALSE</span></span><br><span class="line">  );</span><br><span class="line">&gt; plotFeatureSingle(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    feature.value=log(x.sp@metaData[,<span class="string">"passed_filters"</span>]+<span class="number">1</span>,<span class="number">10</span>),</span><br><span class="line">    method=<span class="string">"tsne"</span>, </span><br><span class="line">    main=<span class="string">"10X Brain Read Depth"</span>,</span><br><span class="line">    point.size=<span class="number">0.2</span>, </span><br><span class="line">    point.shape=<span class="number">19</span>, </span><br><span class="line">    down.sample=<span class="number">10000</span>,</span><br><span class="line">    quantiles=c(<span class="number">0.01</span>, <span class="number">0.99</span>)</span><br><span class="line">  ); </span><br><span class="line">&gt; plotFeatureSingle(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    feature.value=x.sp@metaData$peak_region_fragments / x.sp@metaData$passed_filters,</span><br><span class="line">    method=<span class="string">"tsne"</span>, </span><br><span class="line">    main=<span class="string">"10X Brain FRiP"</span>,</span><br><span class="line">    point.size=<span class="number">0.2</span>, </span><br><span class="line">    point.shape=<span class="number">19</span>, </span><br><span class="line">    down.sample=<span class="number">10000</span>,</span><br><span class="line">    quantiles=c(<span class="number">0.01</span>, <span class="number">0.99</span>) <span class="comment"># remove outliers</span></span><br><span class="line">  );</span><br><span class="line">&gt; plotFeatureSingle(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    feature.value=x.sp@metaData$duplicate / x.sp@metaData$total,</span><br><span class="line">    method=<span class="string">"tsne"</span>, </span><br><span class="line">    main=<span class="string">"10X Brain Duplicate"</span>,</span><br><span class="line">    point.size=<span class="number">0.2</span>, </span><br><span class="line">    point.shape=<span class="number">19</span>, </span><br><span class="line">    down.sample=<span class="number">10000</span>,</span><br><span class="line">    quantiles=c(<span class="number">0.01</span>, <span class="number">0.99</span>) <span class="comment"># remove outliers</span></span><br><span class="line">  );</span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-4-Visulization.jpg" width="900" height="900" />  
<h2 id="a-namegene_tsneastep-9-gene-based-annotation"><a class="markdownIt-Anchor" href="#a-namegene_tsneastep-9-gene-based-annotation"></a> <a name="gene_tsne"></a><strong>Step 9. Gene based annotation</strong></h2>
<p>To help annotate identified cell clusters, SnapATAC next creates the cell-by-gene matrix and visualize the enrichment of marker genes.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; system(<span class="string">"wget http://renlab.sdsc.edu/r3fang/share/github/Mouse_Brain_10X/gencode.vM16.gene.bed"</span>);</span><br><span class="line">&gt; genes = read.table(<span class="string">"gencode.vM16.gene.bed"</span>);</span><br><span class="line">&gt; genes.gr = GRanges(genes[,<span class="number">1</span>], </span><br><span class="line">    IRanges(genes[,<span class="number">2</span>], genes[,<span class="number">3</span>]), name=genes[,<span class="number">4</span>]</span><br><span class="line">  );</span><br><span class="line">&gt; marker.genes = c(</span><br><span class="line">    <span class="string">"Snap25"</span>, <span class="string">"Gad2"</span>, <span class="string">"Apoe"</span>,</span><br><span class="line">    <span class="string">"C1qb"</span>, <span class="string">"Pvalb"</span>, <span class="string">"Vip"</span>, </span><br><span class="line">    <span class="string">"Sst"</span>, <span class="string">"Lamp5"</span>, <span class="string">"Slc17a7"</span></span><br><span class="line">  );</span><br><span class="line">&gt; genes.sel.gr &lt;- genes.gr[which(genes.gr$name %<span class="keyword">in</span>% marker.genes)];</span><br><span class="line"><span class="comment"># re-add the cell-by-bin matrix to the snap object;</span></span><br><span class="line">&gt; x.sp = addBmatToSnap(x.sp);</span><br><span class="line">&gt; x.sp = createGmatFromMat(</span><br><span class="line">    obj=x.sp, </span><br><span class="line">    input.mat=<span class="string">"bmat"</span>,</span><br><span class="line">    genes=genes.sel.gr,</span><br><span class="line">    do.par=<span class="literal">TRUE</span>,</span><br><span class="line">    num.cores=<span class="number">10</span></span><br><span class="line">  );</span><br><span class="line"><span class="comment"># normalize the cell-by-gene matrix</span></span><br><span class="line">&gt; x.sp = scaleCountMatrix(</span><br><span class="line">    obj=x.sp, </span><br><span class="line">    cov=x.sp@metaData$passed_filters + <span class="number">1</span>,</span><br><span class="line">    mat=<span class="string">"gmat"</span>,</span><br><span class="line">    method = <span class="string">"RPM"</span></span><br><span class="line">  );</span><br><span class="line"><span class="comment"># smooth the cell-by-gene matrix</span></span><br><span class="line">&gt; x.sp = runMagic(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    input.mat=<span class="string">"gmat"</span>,</span><br><span class="line">    step.size=<span class="number">3</span></span><br><span class="line">  );</span><br><span class="line">&gt; par(mfrow = c(<span class="number">3</span>, <span class="number">3</span>));</span><br><span class="line">&gt; <span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:<span class="number">9</span>)&#123;</span><br><span class="line">    plotFeatureSingle(</span><br><span class="line">        obj=x.sp,</span><br><span class="line">        feature.value=x.sp@gmat[, marker.genes[i]],</span><br><span class="line">        method=<span class="string">"tsne"</span>, </span><br><span class="line">        main=marker.genes[i],</span><br><span class="line">        point.size=<span class="number">0.1</span>, </span><br><span class="line">        point.shape=<span class="number">19</span>, </span><br><span class="line">        down.sample=<span class="number">10000</span>,</span><br><span class="line">        quantiles=c(<span class="number">0</span>, <span class="number">1</span>)</span><br><span class="line">  )&#125;;</span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-5-gene_plot.jpg" width="900" height="900" />
<h2 id="a-nameheretical_clusteringastep-10-heretical-clustering"><a class="markdownIt-Anchor" href="#a-nameheretical_clusteringastep-10-heretical-clustering"></a> <a name="heretical_clustering"></a><strong>Step 10. Heretical clustering</strong></h2>
<p>Next, cells belonging to the same cluster are pooled to create the aggregate signal for heretical clustering.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment"># calculate the ensemble signals for each cluster</span></span><br><span class="line">&gt; ensemble.ls = lapply(split(seq(length(x.sp@cluster)), x.sp@cluster), <span class="keyword">function</span>(x)&#123;</span><br><span class="line">	SnapATAC::colMeans(x.sp[x,], mat=<span class="string">"bmat"</span>);</span><br><span class="line">	&#125;)</span><br><span class="line"><span class="comment"># cluster using 1-cor as distance  </span></span><br><span class="line">&gt; hc = hclust(as.dist(<span class="number">1</span> - cor(t(do.call(rbind, ensemble.ls)))), method=<span class="string">"ward.D2"</span>);</span><br><span class="line">&gt; plotViz(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    method=<span class="string">"tsne"</span>, </span><br><span class="line">    main=<span class="string">"10X Brain Cluster"</span>,</span><br><span class="line">    point.color=x.sp@cluster, </span><br><span class="line">    point.size=<span class="number">1</span>, </span><br><span class="line">    point.shape=<span class="number">19</span>, </span><br><span class="line">    point.alpha=<span class="number">0.8</span>, </span><br><span class="line">    text.add=<span class="literal">TRUE</span>,</span><br><span class="line">    text.size=<span class="number">1.5</span>,</span><br><span class="line">    text.color=<span class="string">"black"</span>,</span><br><span class="line">    text.halo.add=<span class="literal">TRUE</span>,</span><br><span class="line">    text.halo.color=<span class="string">"white"</span>,</span><br><span class="line">    text.halo.width=<span class="number">0.2</span>,</span><br><span class="line">    down.sample=<span class="number">10000</span>,</span><br><span class="line">    legend.add=<span class="literal">FALSE</span></span><br><span class="line">    );</span><br><span class="line">&gt; plot(hc, hang=-<span class="number">1</span>, xlab=<span class="string">""</span>);</span><br></pre></td></tr></table></figure>
<p>In this case, from cluster 20 to 25 are excitatory neuron; cluster 19 to 5 are inhibitory neurons and the rest of them are non-neuronal cells.</p>
<table opacity=‚Äú0‚Äù rules=none frame=void><tr>
<td bgcolor="white" opacity=‚Äú0‚Äù><img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-6-Visulization_tsne.jpg" width="400" height="400" /></td> 
<td bgcolor="white" opacity=‚Äú0‚Äù><img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-7-cluster_tree.jpg" width="400" height="400" /></td>
</tr></table>
<h2 id="a-namepeak_callastep-11-identify-peaks"><a class="markdownIt-Anchor" href="#a-namepeak_callastep-11-identify-peaks"></a> <a name="peak_call"></a><strong>Step 11. Identify peaks</strong></h2>
<p>Next we aggregate cells from the each cluster to create an ensemble track for peak calling and visualization. This step will generate a .narrowPeak file that contains the identified peak and .bedGraph file for visualization. To obtain the most robust result, we don‚Äôt recommend to perform this step for clusters with cell number less than 100. In the below example, SnapATAC creates <mark class="label default"><font color=c52950>atac_v1_adult_brain_fresh_5k.1_peaks.narrowPeak</font></mark> and <mark class="label default"><font color=c52950>atac_v1_adult_brain_fresh_5k.1_treat_pileup.bdg</font></mark>. bdg file can be compressed to bigWig file using bedGraphToBigWig for IGV or Genome Browser visulization.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; system(<span class="string">"which snaptools"</span>)</span><br><span class="line">/home/r3fang/anaconda2/bin/snaptools</span><br><span class="line">&gt; system(<span class="string">"which macs2"</span>)</span><br><span class="line">/home/r3fang/anaconda2/bin/macs2</span><br><span class="line">&gt; runMACS(</span><br><span class="line">	obj=x.sp[which(x.sp@cluster==<span class="number">1</span>),], </span><br><span class="line">	output.prefix=<span class="string">"atac_v1_adult_brain_fresh_5k.1"</span>,</span><br><span class="line">	path.to.snaptools=<span class="string">"/home/r3fang/anaconda2/bin/snaptools"</span>,</span><br><span class="line">	path.to.macs=<span class="string">"/home/r3fang/anaconda2/bin/macs2"</span>,</span><br><span class="line">	gsize=<span class="string">"mm"</span>, </span><br><span class="line">	buffer.size=<span class="number">500</span>, </span><br><span class="line">	num.cores=<span class="number">5</span>,</span><br><span class="line">	macs.options=<span class="string">"--nomodel --shift 37 --ext 73 --qval 1e-2 -B --SPMR --call-summits"</span>,</span><br><span class="line">	tmp.folder=tempdir()</span><br><span class="line">	);</span><br></pre></td></tr></table></figure>
<p>Next, we provide a short script that performs this step for all clusters.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment"># call peaks for all cluster with more than 100 cells</span></span><br><span class="line">&gt; clusters.sel = names(table(x.sp@cluster))[which(table(x.sp@cluster) &gt; <span class="number">200</span>)];</span><br><span class="line">&gt; peaks.ls = mclapply(seq(clusters.sel), <span class="keyword">function</span>(i)&#123;</span><br><span class="line">    print(clusters.sel[i]);</span><br><span class="line">    runMACS(</span><br><span class="line">        obj=x.sp[which(x.sp@cluster==clusters.sel[i]),], </span><br><span class="line">        output.prefix=paste0(<span class="string">"atac_v1_adult_brain_fresh_5k."</span>, gsub(<span class="string">" "</span>, <span class="string">"_"</span>, clusters.sel)[i]),</span><br><span class="line">        path.to.snaptools=<span class="string">"/home/r3fang/anaconda2/bin/snaptools"</span>,</span><br><span class="line">        path.to.macs=<span class="string">"/home/r3fang/anaconda2/bin/macs2"</span>,</span><br><span class="line">        gsize=<span class="string">"hs"</span>, <span class="comment"># mm, hs, etc</span></span><br><span class="line">        buffer.size=<span class="number">500</span>, </span><br><span class="line">        num.cores=<span class="number">1</span>,</span><br><span class="line">        macs.options=<span class="string">"--nomodel --shift 100 --ext 200 --qval 5e-2 -B --SPMR"</span>,</span><br><span class="line">        tmp.folder=tempdir()</span><br><span class="line">   );</span><br><span class="line"> &#125;, mc.cores=<span class="number">5</span>);</span><br><span class="line"><span class="comment"># assuming all .narrowPeak files in the current folder are generated from the clusters</span></span><br><span class="line">&gt; peaks.names = system(<span class="string">"ls | grep narrowPeak"</span>, intern=<span class="literal">TRUE</span>);</span><br><span class="line">&gt; peak.gr.ls = lapply(peaks.names, <span class="keyword">function</span>(x)&#123;</span><br><span class="line">    peak.df = read.table(x)</span><br><span class="line">    GRanges(peak.df[,<span class="number">1</span>], IRanges(peak.df[,<span class="number">2</span>], peak.df[,<span class="number">3</span>]))</span><br><span class="line">  &#125;)</span><br><span class="line">&gt; peak.gr = reduce(Reduce(c, peak.gr.ls));</span><br><span class="line">&gt; peak.gr</span><br></pre></td></tr></table></figure>
<figure class="highlight angelscript"><table><tr><td class="code"><pre><span class="line">## GRanges object with <span class="number">242847</span> ranges <span class="keyword">and</span> <span class="number">0</span> metadata columns:</span><br><span class="line">##           seqnames               ranges strand</span><br><span class="line">##              &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;</span><br><span class="line">##       [<span class="number">1</span>]     chr1   [<span class="number">3094889</span>, <span class="number">3095629</span>]      *</span><br><span class="line">##       [<span class="number">2</span>]     chr1   [<span class="number">3113499</span>, <span class="number">3114060</span>]      *</span><br><span class="line">##       [<span class="number">3</span>]     chr1   [<span class="number">3118103</span>, <span class="number">3118401</span>]      *</span><br><span class="line">##       [<span class="number">4</span>]     chr1   [<span class="number">3119689</span>, <span class="number">3120845</span>]      *</span><br><span class="line">##       [<span class="number">5</span>]     chr1   [<span class="number">3121534</span>, <span class="number">3121786</span>]      *</span><br><span class="line">##       ...      ...                  ...    ...</span><br><span class="line">##  [<span class="number">242843</span>]     chrY [<span class="number">90797373</span>, <span class="number">90798136</span>]      *</span><br><span class="line">##  [<span class="number">242844</span>]     chrY [<span class="number">90804709</span>, <span class="number">90805456</span>]      *</span><br><span class="line">##  [<span class="number">242845</span>]     chrY [<span class="number">90808580</span>, <span class="number">90808819</span>]      *</span><br><span class="line">##  [<span class="number">242846</span>]     chrY [<span class="number">90808850</span>, <span class="number">90809131</span>]      *</span><br><span class="line">##  [<span class="number">242847</span>]     chrY [<span class="number">90810817</span>, <span class="number">90811057</span>]      *</span><br><span class="line">##  -------</span><br></pre></td></tr></table></figure>
<h2 id="a-namecreate_pmatastep-12-create-a-cell-by-peak-matrix"><a class="markdownIt-Anchor" href="#a-namecreate_pmatastep-12-create-a-cell-by-peak-matrix"></a> <a name="create_pmat"></a><strong>Step 12. Create a cell-by-peak matrix</strong></h2>
<p>Using merged peak list as a reference, we next create a cell-by-peak matrix using the original snap file.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; peaks.df = as.data.frame(peak.gr)[,<span class="number">1</span>:<span class="number">3</span>];</span><br><span class="line">&gt; write.table(peaks.df,file = <span class="string">"peaks.combined.bed"</span>,append=<span class="literal">FALSE</span>,</span><br><span class="line">		quote= <span class="literal">FALSE</span>,sep=<span class="string">"\t"</span>, eol = <span class="string">"\n"</span>, na = <span class="string">"NA"</span>, dec = <span class="string">"."</span>, </span><br><span class="line">		row.names = <span class="literal">FALSE</span>, col.names = <span class="literal">FALSE</span>, qmethod = c(<span class="string">"escape"</span>, <span class="string">"double"</span>),</span><br><span class="line">		fileEncoding = <span class="string">""</span>)</span><br><span class="line">&gt; saveRDS(x.sp, file=<span class="string">"atac_v1_adult_brain_fresh_5k.snap.rds"</span>);</span><br></pre></td></tr></table></figure>
<p>Next we create cell-by-peak matrix and add to the snap file. This step will take a while.</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">$ snaptools snap-add-pmat \</span><br><span class="line">	--snap-file atac_v1_adult_brain_fresh_5k.snap \</span><br><span class="line">	--peak-file peaks.combined.bed</span><br></pre></td></tr></table></figure>
<h2 id="a-nameadd_pmatastep-13-add-cell-by-peak-matrix"><a class="markdownIt-Anchor" href="#a-nameadd_pmatastep-13-add-cell-by-peak-matrix"></a> <a name="add_pmat"></a><strong>Step 13. Add cell-by-peak matrix</strong></h2>
<p>Next we add the cell-by-peak matrix to the existing snap object.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; x.sp = readRDS(<span class="string">"atac_v1_adult_brain_fresh_5k.snap.rds"</span>);</span><br><span class="line">&gt; x.sp = addPmatToSnap(x.sp);</span><br><span class="line">&gt; x.sp = makeBinary(x.sp, mat=<span class="string">"pmat"</span>);</span><br><span class="line">&gt; x.sp</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## number <span class="keyword">of</span> barcodes: <span class="number">4100</span></span><br><span class="line">## number <span class="keyword">of</span> bins: <span class="number">546206</span></span><br><span class="line">## number <span class="keyword">of</span> genes: <span class="number">16</span></span><br><span class="line">## number <span class="keyword">of</span> peaks: <span class="number">242847</span></span><br><span class="line">## number <span class="keyword">of</span> motifs: <span class="number">0</span></span><br></pre></td></tr></table></figure>
<h2 id="a-namediff_analysisastep-14-identify-differentially-accessible-regions"><a class="markdownIt-Anchor" href="#a-namediff_analysisastep-14-identify-differentially-accessible-regions"></a> <a name="diff_analysis"></a><strong>Step 14. Identify differentially accessible regions</strong></h2>
<p>SnapATAC finds differentially accessible regions (DARs) that define clusters via differential analysis. By default, it identifies positive peaks of a single cluster (specified in <mark class="label default"><font color=c52950>cluster.pos</font></mark>), compared to a group of negative control cells. If by default <mark class="label default"><font color=c52950>cluster.neg=NULL</font></mark>, findDAR will look for a group of background cells closest to the positive cells.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; DARs = findDAR(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    input.mat=<span class="string">"pmat"</span>,</span><br><span class="line">    cluster.pos=<span class="number">26</span>,</span><br><span class="line">    cluster.neg.method=<span class="string">"knn"</span>,</span><br><span class="line">    test.method=<span class="string">"exactTest"</span>,</span><br><span class="line">    bcv=<span class="number">0.1</span>, <span class="comment">#0.4 for human, 0.1 for mouse</span></span><br><span class="line">    seed.use=<span class="number">10</span></span><br><span class="line">  );</span><br><span class="line">&gt; DARs$FDR = p.adjust(DARs$PValue, method=<span class="string">"BH"</span>);</span><br><span class="line">&gt; idy = which(DARs$FDR &lt; <span class="number">5e-2</span> &amp; DARs$logFC &gt; <span class="number">0</span>);</span><br><span class="line">&gt; par(mfrow = c(<span class="number">1</span>, <span class="number">2</span>));</span><br><span class="line">&gt; plot(DARs$logCPM, DARs$logFC, </span><br><span class="line">    pch=<span class="number">19</span>, cex=<span class="number">0.1</span>, col=<span class="string">"grey"</span>, </span><br><span class="line">    ylab=<span class="string">"logFC"</span>, xlab=<span class="string">"logCPM"</span>,</span><br><span class="line">    main=<span class="string">"Cluster 26"</span></span><br><span class="line">  );</span><br><span class="line">&gt; points(DARs$logCPM[idy], </span><br><span class="line">    DARs$logFC[idy], </span><br><span class="line">    pch=<span class="number">19</span>, </span><br><span class="line">    cex=<span class="number">0.5</span>, </span><br><span class="line">    col=<span class="string">"red"</span></span><br><span class="line">  );</span><br><span class="line">&gt; abline(h = <span class="number">0</span>, lwd=<span class="number">1</span>, lty=<span class="number">2</span>);</span><br><span class="line">&gt; covs = Matrix::rowSums(x.sp@pmat);</span><br><span class="line">&gt; vals = Matrix::rowSums(x.sp@pmat[,idy]) / covs;</span><br><span class="line">&gt; vals.zscore = (vals - mean(vals)) / sd(vals);</span><br><span class="line">&gt; plotFeatureSingle(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    feature.value=vals.zscore,</span><br><span class="line">    method=<span class="string">"tsne"</span>, </span><br><span class="line">    main=<span class="string">"Cluster 26"</span>,</span><br><span class="line">    point.size=<span class="number">0.1</span>, </span><br><span class="line">    point.shape=<span class="number">19</span>, </span><br><span class="line">    down.sample=<span class="number">5000</span>,</span><br><span class="line">    quantiles=c(<span class="number">0.01</span>, <span class="number">0.99</span>)</span><br><span class="line">  );</span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-8-DARs.jpg" width="700" height="350" /> 
<p>Next, we identify DARs for each of the clusters. For clusters, especially the small ones, that lack the static power to reveal DARs, we rank the peaks based on the enrichment and use the top 2000 peaks as representative peaks for motif discovery.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; idy.ls = lapply(levels(x.sp@cluster), <span class="keyword">function</span>(cluster_i)&#123;</span><br><span class="line">	DARs = findDAR(</span><br><span class="line">		obj=x.sp,</span><br><span class="line">		input.mat=<span class="string">"pmat"</span>,</span><br><span class="line">		cluster.pos=cluster_i,</span><br><span class="line">		cluster.neg=<span class="literal">NULL</span>,</span><br><span class="line">		cluster.neg.method=<span class="string">"knn"</span>,</span><br><span class="line">		bcv=<span class="number">0.1</span>,</span><br><span class="line">		test.method=<span class="string">"exactTest"</span>,</span><br><span class="line">		seed.use=<span class="number">10</span></span><br><span class="line">		);</span><br><span class="line">	DARs$FDR = p.adjust(DARs$PValue, method=<span class="string">"BH"</span>);</span><br><span class="line">	idy = which(DARs$FDR &lt; <span class="number">5e-2</span> &amp; DARs$logFC &gt; <span class="number">0</span>);</span><br><span class="line">	<span class="keyword">if</span>((x=length(idy)) &lt; <span class="number">2000L</span>)&#123;</span><br><span class="line">			PValues = DARs$PValue;</span><br><span class="line">			PValues[DARs$logFC &lt; <span class="number">0</span>] = <span class="number">1</span>;</span><br><span class="line">			idy = order(PValues, decreasing=<span class="literal">FALSE</span>)[<span class="number">1</span>:<span class="number">2000</span>];</span><br><span class="line">			rm(PValues); <span class="comment"># free memory</span></span><br><span class="line">	&#125;</span><br><span class="line">	idy</span><br><span class="line">  &#125;)</span><br><span class="line">&gt; names(idy.ls) = levels(x.sp@cluster);</span><br><span class="line">&gt; par(mfrow = c(<span class="number">3</span>, <span class="number">3</span>));</span><br><span class="line">&gt; <span class="keyword">for</span>(cluster_i <span class="keyword">in</span> levels(x.sp@cluster))&#123;</span><br><span class="line">	print(cluster_i)</span><br><span class="line">	idy = idy.ls[[cluster_i]];</span><br><span class="line">	vals = Matrix::rowSums(x.sp@pmat[,idy]) / covs;</span><br><span class="line">	vals.zscore = (vals - mean(vals)) / sd(vals);</span><br><span class="line">	plotFeatureSingle(</span><br><span class="line">		obj=x.sp,</span><br><span class="line">		feature.value=vals.zscore,</span><br><span class="line">		method=<span class="string">"tsne"</span>, </span><br><span class="line">		main=cluster_i,</span><br><span class="line">		point.size=<span class="number">0.1</span>, </span><br><span class="line">		point.shape=<span class="number">19</span>, </span><br><span class="line">		down.sample=<span class="number">5000</span>,</span><br><span class="line">		quantiles=c(<span class="number">0.01</span>, <span class="number">0.99</span>)</span><br><span class="line">		);</span><br><span class="line">  &#125;</span><br></pre></td></tr></table></figure>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-9-DARs_all.jpg" width="900" height="900" /> 
<h2 id="a-namehomer_chromvarastep-15-motif-analysis-identifies-master-regulators"><a class="markdownIt-Anchor" href="#a-namehomer_chromvarastep-15-motif-analysis-identifies-master-regulators"></a> <a name="homer_chromVAR"></a><strong>Step 15. Motif analysis identifies master regulators</strong></h2>
<p>SnapATAC employs Homer to identify master regulators that are enriched in the differentially accessible regions (DARs). This will creates a homer motif report <mark class="label default"><font color=c52950>knownResults.html</font></mark> in the folder <mark class="label default"><font color=c52950>./homer/C5</font></mark>. This requires Homer to be pre-installed. See <span class="exturl" data-url="aHR0cDovL2hvbWVyLnVjc2QuZWR1L2hvbWVyL2ludHJvZHVjdGlvbi9pbnN0YWxsLmh0bWw=">here<i class="fa fa-external-link-alt"></i></span> for the instruction about how to install Homer.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; system(<span class="string">"which findMotifsGenome.pl"</span>);</span><br><span class="line">/projects/ps-renlab/r3fang/public_html/softwares/homer/bin/findMotifsGenome.pl</span><br><span class="line">&gt; motifs = runHomer(</span><br><span class="line">	x.sp[,idy.ls[[<span class="string">"5"</span>]],<span class="string">"pmat"</span>], </span><br><span class="line">	mat = <span class="string">"pmat"</span>,</span><br><span class="line">	path.to.homer = <span class="string">"/projects/ps-renlab/r3fang/public_html/softwares/homer/bin/findMotifsGenome.pl"</span>,</span><br><span class="line">	result.dir = <span class="string">"./homer/C5"</span>,</span><br><span class="line">	num.cores=<span class="number">5</span>,</span><br><span class="line">	genome = <span class="string">'mm10'</span>,</span><br><span class="line">	motif.length = <span class="number">10</span>,</span><br><span class="line">	scan.size = <span class="number">300</span>,</span><br><span class="line">	optimize.count = <span class="number">2</span>,</span><br><span class="line">	background = <span class="string">'automatic'</span>,</span><br><span class="line">	local.background = <span class="literal">FALSE</span>,</span><br><span class="line">	only.known = <span class="literal">TRUE</span>,</span><br><span class="line">	only.denovo = <span class="literal">FALSE</span>,</span><br><span class="line">	fdr.num = <span class="number">5</span>,</span><br><span class="line">	cache = <span class="number">100</span>,</span><br><span class="line">	overwrite = <span class="literal">TRUE</span>,</span><br><span class="line">	keep.minimal = <span class="literal">FALSE</span></span><br><span class="line">  );</span><br></pre></td></tr></table></figure>
<p>See <span class="exturl" data-url="aHR0cDovL3JlbmxhYi5zZHNjLmVkdS9yM2Zhbmcvc2hhcmUvZ2l0aHViL01vdXNlX0JyYWluXzEwWC9ob21lci9DNS9rbm93blJlc3VsdHMuaHRtbA==">here<i class="fa fa-external-link-alt"></i></span> for the full list of motifs for cluster 5.</p>
<img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-10-homer_motif.jpg" width="900" height="400" /> 
<p>SnapATAC also incorporates chromVAR (Schep et al) for motif variability analysis.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; <span class="keyword">library</span>(chromVAR);</span><br><span class="line">&gt; <span class="keyword">library</span>(motifmatchr);</span><br><span class="line">&gt; <span class="keyword">library</span>(SummarizedExperiment);</span><br><span class="line">&gt; <span class="keyword">library</span>(BSgenome.Mmusculus.UCSC.mm10);</span><br><span class="line">&gt; x.sp = makeBinary(x.sp, <span class="string">"pmat"</span>);</span><br><span class="line">&gt; x.sp@mmat = runChromVAR(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    input.mat=<span class="string">"pmat"</span>,</span><br><span class="line">    genome=BSgenome.Mmusculus.UCSC.mm10,</span><br><span class="line">    min.count=<span class="number">10</span>,</span><br><span class="line">    species=<span class="string">"Homo sapiens"</span></span><br><span class="line">  );</span><br><span class="line">&gt; motif_i = <span class="string">"MA0497.1_MEF2C"</span>;</span><br><span class="line">&gt; dat = data.frame(x=x.sp@metaData[,<span class="string">"cluster"</span>], y=x.sp@mmat[,motif_i]);</span><br><span class="line">&gt; p1 &lt;- ggplot(dat, aes(x=x, y=y, fill=x)) + </span><br><span class="line">	theme_classic() +</span><br><span class="line">	geom_violin() + </span><br><span class="line">	xlab(<span class="string">"cluster"</span>) +</span><br><span class="line">	ylab(<span class="string">"motif enrichment"</span>) + </span><br><span class="line">	ggtitle(motif_i) +</span><br><span class="line">	theme(</span><br><span class="line">		  plot.margin = margin(<span class="number">5</span>,<span class="number">1</span>,<span class="number">5</span>,<span class="number">1</span>, <span class="string">"cm"</span>),</span><br><span class="line">		  axis.text.x = element_text(angle = <span class="number">90</span>, hjust = <span class="number">1</span>),</span><br><span class="line">		  axis.ticks.x=element_blank(),</span><br><span class="line">		  legend.position = <span class="string">"none"</span></span><br><span class="line">   );</span><br><span class="line">&gt; motif_i = <span class="string">"MA0660.1_MEF2B"</span>;</span><br><span class="line">&gt; dat = data.frame(x=x.sp@metaData[,<span class="string">"cluster"</span>], y=x.sp@mmat[,motif_i]);</span><br><span class="line">&gt; p2 &lt;- ggplot(dat, aes(x=x, y=y, fill=x)) + </span><br><span class="line">	theme_classic() +</span><br><span class="line">	geom_violin() + </span><br><span class="line">	xlab(<span class="string">"cluster"</span>) +</span><br><span class="line">	ylab(<span class="string">"motif enrichment"</span>) + </span><br><span class="line">	ggtitle(motif_i) +</span><br><span class="line">	theme(</span><br><span class="line">		  plot.margin = margin(<span class="number">5</span>,<span class="number">1</span>,<span class="number">5</span>,<span class="number">1</span>, <span class="string">"cm"</span>),</span><br><span class="line">		  axis.text.x = element_text(angle = <span class="number">90</span>, hjust = <span class="number">1</span>),</span><br><span class="line">		  axis.ticks.x=element_blank(),</span><br><span class="line">		  legend.position = <span class="string">"none"</span></span><br><span class="line">   );</span><br><span class="line">&gt; p1</span><br><span class="line">&gt; p2</span><br></pre></td></tr></table></figure>
<table opacity=‚Äú0‚Äù rules=none frame=void><tr>
<td bgcolor="white" opacity=‚Äú0‚Äù><img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-11-chromVAR_2B.jpg" width="900" height="400" /></td>
<td bgcolor="white" opacity=‚Äú0‚Äù><img src="https://cdn.jsdelivr.net/gh/geng-lee/ImageHosting/img/scrna2-12-chromVAR_2C.jpg" width="900" height="400" /></td>
</tr></table>
<h2 id="a-namegreat_analysisastep-16-great-analysis"><a class="markdownIt-Anchor" href="#a-namegreat_analysisastep-16-great-analysis"></a> <a name="great_analysis"></a><strong>Step 16. GREAT analysis</strong></h2>
<p>SnapATAC applies GREAT to identify biological pathways active in each of the cell cluster. In this example, we will first identify the differential elements in Microglia cells (cluster 13) and report the top 6 GO Ontologies enrichment inferred using GREAT analysis.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment">## install R package rGREAT</span></span><br><span class="line">&gt; <span class="keyword">if</span> (!requireNamespace(<span class="string">"BiocManager"</span>, quietly=<span class="literal">TRUE</span>))</span><br><span class="line">    install.packages(<span class="string">"BiocManager"</span>)</span><br><span class="line">&gt; BiocManager::install(<span class="string">"rGREAT"</span>)</span><br><span class="line"><span class="comment">## or install the latest version</span></span><br><span class="line">&gt; <span class="keyword">library</span>(devtools)</span><br><span class="line">&gt; install_github(<span class="string">"jokergoo/rGREAT"</span>)</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; <span class="keyword">library</span>(rGREAT);</span><br><span class="line">&gt; DARs = findDAR(</span><br><span class="line">    obj=x.sp,</span><br><span class="line">    input.mat=<span class="string">"pmat"</span>,</span><br><span class="line">    cluster.pos=<span class="number">13</span>,</span><br><span class="line">    cluster.neg.method=<span class="string">"knn"</span>,</span><br><span class="line">    test.method=<span class="string">"exactTest"</span>,</span><br><span class="line">    bcv=<span class="number">0.1</span>, <span class="comment">#0.4 for human, 0.1 for mouse</span></span><br><span class="line">    seed.use=<span class="number">10</span></span><br><span class="line">  );</span><br><span class="line">&gt; DARs$FDR = p.adjust(DARs$PValue, method=<span class="string">"BH"</span>);</span><br><span class="line">&gt; idy = which(DARs$FDR &lt; <span class="number">5e-2</span> &amp; DARs$logFC &gt; <span class="number">0</span>);</span><br><span class="line">&gt; job = submitGreatJob(</span><br><span class="line">    gr                    = x.sp@peak[idy],</span><br><span class="line">    bg                    = <span class="literal">NULL</span>,</span><br><span class="line">    species               = <span class="string">"mm10"</span>,</span><br><span class="line">    includeCuratedRegDoms = <span class="literal">TRUE</span>,</span><br><span class="line">    rule                  = <span class="string">"basalPlusExt"</span>,</span><br><span class="line">    adv_upstream          = <span class="number">5.0</span>,</span><br><span class="line">    adv_downstream        = <span class="number">1.0</span>,</span><br><span class="line">    adv_span              = <span class="number">1000.0</span>,</span><br><span class="line">    adv_twoDistance       = <span class="number">1000.0</span>,</span><br><span class="line">    adv_oneDistance       = <span class="number">1000.0</span>,</span><br><span class="line">    request_interval = <span class="number">300</span>,</span><br><span class="line">    max_tries = <span class="number">10</span>,</span><br><span class="line">    version = <span class="string">"default"</span>,</span><br><span class="line">    base_url = <span class="string">"http://great.stanford.edu/public/cgi-bin"</span></span><br><span class="line">  );</span><br><span class="line">&gt; job</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## Submit time: <span class="number">2019</span><span class="number">-09</span><span class="number">-04</span> <span class="number">14</span>:<span class="number">14</span>:<span class="number">02</span></span><br><span class="line">## Version: default</span><br><span class="line">## Species: mm10</span><br><span class="line">## Inputs: <span class="number">25120</span> regions</span><br><span class="line">## Background: wholeGenome</span><br><span class="line">## Model: Basal plus extension</span><br><span class="line">##   Proximal: <span class="number">5</span> kb upstream, <span class="number">1</span> kb downstream,</span><br><span class="line">##   plus Distal: up to <span class="number">1000</span> kb</span><br><span class="line">## Include curated regulatory domains</span><br><span class="line">## </span><br><span class="line">## Enrichment tables for following ontologies have been downloaded:</span><br><span class="line">##   None</span><br></pre></td></tr></table></figure>
<p>With <code>job</code>, we can now retrieve results from GREAT. The first and the primary results are the tables which contain enrichment statistics for the analysis. By default it will retrieve results from three GO Ontologies and all pathway ontologies. All tables contains statistics for all terms no matter they are significant or not. Users can then make filtering by a self-defined cutoff.</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">&gt; tb = getEnrichmentTables(job);</span><br><span class="line">&gt; names(tb);</span><br></pre></td></tr></table></figure>
<figure class="highlight clean"><table><tr><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="string">"GO Molecular Function"</span> <span class="string">"GO Biological Process"</span> <span class="string">"GO Cellular Component"</span></span><br></pre></td></tr></table></figure>
<figure class="highlight lua"><table><tr><td class="code"><pre><span class="line">&gt; GBP = tb<span class="string">[["GO Biological Process"]]</span>;</span><br><span class="line">&gt; head(GBP[order(GBP[,<span class="string">"Binom_Adjp_BH"</span>]),<span class="number">1</span>:<span class="number">5</span>]);</span><br></pre></td></tr></table></figure>
<figure class="highlight angelscript"><table><tr><td class="code"><pre><span class="line">##           ID                                      name Binom_Genome_Fraction</span><br><span class="line">## <span class="number">1</span> GO:<span class="number">0002376</span>                     immune system process            <span class="number">0.12515840</span></span><br><span class="line">## <span class="number">2</span> GO:<span class="number">0002682</span>       regulation of immune system process            <span class="number">0.09012561</span></span><br><span class="line">## <span class="number">3</span> GO:<span class="number">0009987</span>                          cellular process            <span class="number">0.80870120</span></span><br><span class="line">## <span class="number">4</span> GO:<span class="number">0048518</span> positive regulation of biological process            <span class="number">0.43002240</span></span><br><span class="line">## <span class="number">5</span> GO:<span class="number">0050789</span>          regulation of biological process            <span class="number">0.68873070</span></span><br><span class="line">## <span class="number">6</span> GO:<span class="number">0050794</span>            regulation of cellular process            <span class="number">0.66837300</span></span><br><span class="line">##   Binom_Expected Binom_Observed_Region_Hits</span><br><span class="line">## <span class="number">1</span>       <span class="number">3095.918</span>                       <span class="number">5592</span></span><br><span class="line">## <span class="number">2</span>       <span class="number">2229.347</span>                       <span class="number">4148</span></span><br><span class="line">## <span class="number">3</span>      <span class="number">20004.030</span>                      <span class="number">22241</span></span><br><span class="line">## <span class="number">4</span>      <span class="number">10637.030</span>                      <span class="number">13697</span></span><br><span class="line">## <span class="number">5</span>      <span class="number">17036.440</span>                      <span class="number">19871</span></span><br><span class="line">## <span class="number">6</span>      <span class="number">16532.870</span>                      <span class="number">19356</span></span><br></pre></td></tr></table></figure>

    </div>

    
    
    

<div>
  
    <div>
    
        <div style="text-align:center;color: #555;font-size:14px;">-------------The End-------------</div>
    
</div>

  
</div>
        <div class="reward-container">
  <div>Buy Me A Coffee</div>
  <button onclick="var qr = document.getElementById('qr'); qr.style.display = (qr.style.display === 'none') ? 'block' : 'none';">
    Donate
  </button>
  <div id="qr" style="display: none;">
      
      <div style="display: inline-block;">
        <img src="/images/wechatpay.jpg" alt="Geng Lee WeChat Pay">
        <p>WeChat Pay</p>
      </div>
      
      <div style="display: inline-block;">
        <img src="/images/alipay.png" alt="Geng Lee Alipay">
        <p>Alipay</p>
      </div>

  </div>
</div>


      <footer class="post-footer">
          <div>
            
             <div>
    
        <center><!-- Go to www.addthis.com/dashboard to customize your tools --> <div class="addthis_inline_share_toolbox_ew9r"></div></center>
    
</div>

            
          </div>
          <div class="post-tags">
              <a href="/tags/R/" rel="tag"><i class="fa fa-tag"></i> R</a>
              <a href="/tags/scRNA/" rel="tag"><i class="fa fa-tag"></i> scRNA</a>
              <a href="/tags/ATAC/" rel="tag"><i class="fa fa-tag"></i> ATAC</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/posts/539d6d7a/" rel="prev" title="Calculation of KaKs Value">
      <i class="fa fa-chevron-left"></i> Calculation of KaKs Value
    </a></div>
      <div class="post-nav-item">
    <a href="/posts/1b8fd861/" rel="next" title="Analysis of single-cell ATAC-seq data using SnapATAC software (3)ÔºöIntegrative Analysis of PBMC scATAC-seq and scRNA-seq">
      Analysis of single-cell ATAC-seq data using SnapATAC software (3)ÔºöIntegrative Analysis of PBMC scATAC-seq and scRNA-seq <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  
  


          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#introduction"><span class="nav-number">1.</span> <span class="nav-text"> Introduction</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#10x-adult-mouse-brain"><span class="nav-number">1.1.</span> <span class="nav-text"> 10X Adult Mouse Brain</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#analysis-content"><span class="nav-number">2.</span> <span class="nav-text"> Analysis content</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#table-of-contents"><span class="nav-number">2.1.</span> <span class="nav-text"> Table of Contents</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namebarcode_selectionastep-1-barcode-selection"><span class="nav-number">2.2.</span> <span class="nav-text"> Step 1. Barcode selection</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-nameadd_bmatastep-2-add-cell-by-bin-matrix"><span class="nav-number">2.3.</span> <span class="nav-text"> Step 2. Add cell-by-bin matrix</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namemake_binaryastep-3-matrix-binarization"><span class="nav-number">2.4.</span> <span class="nav-text"> Step 3. Matrix binarization</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namebin_filterastep-4-bin-filtering"><span class="nav-number">2.5.</span> <span class="nav-text"> Step 4. Bin filtering</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namediffusion_mapsastep-5-dimensionality-reduction"><span class="nav-number">2.6.</span> <span class="nav-text"> Step 5. Dimensionality reduction</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namepc_selectastep-6-determine-significant-components"><span class="nav-number">2.7.</span> <span class="nav-text"> Step 6. Determine significant components</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-nameclusterastep-7-graph-based-clustering"><span class="nav-number">2.8.</span> <span class="nav-text"> Step 7. Graph-based clustering</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namevizastep-8-visualization"><span class="nav-number">2.9.</span> <span class="nav-text"> Step 8. Visualization</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namegene_tsneastep-9-gene-based-annotation"><span class="nav-number">2.10.</span> <span class="nav-text"> Step 9. Gene based annotation</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-nameheretical_clusteringastep-10-heretical-clustering"><span class="nav-number">2.11.</span> <span class="nav-text"> Step 10. Heretical clustering</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namepeak_callastep-11-identify-peaks"><span class="nav-number">2.12.</span> <span class="nav-text"> Step 11. Identify peaks</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namecreate_pmatastep-12-create-a-cell-by-peak-matrix"><span class="nav-number">2.13.</span> <span class="nav-text"> Step 12. Create a cell-by-peak matrix</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-nameadd_pmatastep-13-add-cell-by-peak-matrix"><span class="nav-number">2.14.</span> <span class="nav-text"> Step 13. Add cell-by-peak matrix</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namediff_analysisastep-14-identify-differentially-accessible-regions"><span class="nav-number">2.15.</span> <span class="nav-text"> Step 14. Identify differentially accessible regions</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namehomer_chromvarastep-15-motif-analysis-identifies-master-regulators"><span class="nav-number">2.16.</span> <span class="nav-text"> Step 15. Motif analysis identifies master regulators</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#a-namegreat_analysisastep-16-great-analysis"><span class="nav-number">2.17.</span> <span class="nav-text"> Step 16. GREAT analysis</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Geng Lee"
      src="/images/avatar.jpg">
  <p class="site-author-name" itemprop="name">Geng Lee</p>
  <div class="site-description" itemprop="description">üòè To be a better me</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">10</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">6</span>
        <span class="site-state-item-name">categories</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">14</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <span class="exturl" data-url="aHR0cHM6Ly9naXRodWIuY29tL2dlbmctbGVl" title="GitHub ‚Üí https:&#x2F;&#x2F;github.com&#x2F;geng-lee"><i class="fab fa-github fa-fw"></i>GitHub</span>
      </span>
      <span class="links-of-author-item">
        <span class="exturl" data-url="aHR0cHM6Ly93d3cuZmFjZWJvb2suY29tL3Byb2ZpbGUucGhwP2lkPTEwMDAxMTQxODkzNDg5Mg==" title="Facebook ‚Üí https:&#x2F;&#x2F;www.facebook.com&#x2F;profile.php?id&#x3D;100011418934892"><i class="fab fa-facebook fa-fw"></i>Facebook</span>
      </span>
      <span class="links-of-author-item">
        <span class="exturl" data-url="bWFpbHRvOjI3MTY1ODE4NTJAcXEuY29t" title="E-Mail ‚Üí mailto:2716581852@qq.com"><i class="fa fa-envelope fa-fw"></i>E-Mail</span>
      </span>
      <span class="links-of-author-item">
        <span class="exturl" data-url="aHR0cHM6Ly93ZWliby5jb20vdS81MDgwNDg1MzEw" title="Weibo ‚Üí https:&#x2F;&#x2F;weibo.com&#x2F;u&#x2F;5080485310"><i class="fab fa-weibo fa-fw"></i>Weibo</span>
      </span>
  </div>
  <div class="cc-license motion-element" itemprop="license">
    <span class="exturl cc-opacity" data-url="aHR0cHM6Ly9jcmVhdGl2ZWNvbW1vbnMub3JnL2xpY2Vuc2VzL2J5LW5jLXNhLzQuMC8="><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"></span>
  </div>


  <div class="links-of-blogroll motion-element">
    <div class="links-of-blogroll-title"><i class="fa fa-coffee fa-fw"></i>
      
    </div>
    <ul class="links-of-blogroll-list">
        <li class="links-of-blogroll-item">
          <span class="exturl" data-url="aHR0cHM6Ly93d3cuYnV5bWVhY29mZmVlLmNvbS9HZW5nTGVl" title="https:&#x2F;&#x2F;www.buymeacoffee.com&#x2F;GengLee">Buy Me A Coffee</span>
        </li>
    </ul>
  </div>

      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2021</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Geng Lee</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-chart-area"></i>
    </span>
    <span title="Symbols count total">66k</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-coffee"></i>
    </span>
    <span title="Reading time total">1:01</span>
</div>
  <div class="addthis_inline_share_toolbox">
    <script src="//s7.addthis.com/js/300/addthis_widget.js#pubid=ra-5eef5a9cd125218a" async="async"></script>
  </div>

        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="//cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js"></script>
  <script src="//cdn.jsdelivr.net/gh/fancyapps/fancybox@3/dist/jquery.fancybox.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/muse.js"></script>


<script src="/js/next-boot.js"></script>


  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      bp.src = (curProtocol === 'https') ? 'https://zz.bdstatic.com/linksubmit/push.js' : 'http://push.zhanzhang.baidu.com/push.js';
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>




  
<script src="/js/local-search.js"></script>











<script>
if (document.querySelectorAll('pre.mermaid').length) {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/mermaid@8/dist/mermaid.min.js', () => {
    mermaid.initialize({
      theme    : 'forest',
      logLevel : 3,
      flowchart: { curve     : 'linear' },
      gantt    : { axisFormat: '%m/%d/%Y' },
      sequence : { actorMargin: 50 }
    });
  }, window.mermaid);
}
</script>


  

  
      

<script>
  if (typeof MathJax === 'undefined') {
    window.MathJax = {
      loader: {
        source: {
          '[tex]/amsCd': '[tex]/amscd',
          '[tex]/AMScd': '[tex]/amscd'
        }
      },
      tex: {
        inlineMath: {'[+]': [['$', '$']]},
        tags: 'ams'
      },
      options: {
        renderActions: {
          findScript: [10, doc => {
            document.querySelectorAll('script[type^="math/tex"]').forEach(node => {
              const display = !!node.type.match(/; *mode=display/);
              const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display);
              const text = document.createTextNode('');
              node.parentNode.replaceChild(text, node);
              math.start = {node: text, delim: '', n: 0};
              math.end = {node: text, delim: '', n: 0};
              doc.math.push(math);
            });
          }, '', false],
          insertedScript: [200, () => {
            document.querySelectorAll('mjx-container').forEach(node => {
              let target = node.parentNode;
              if (target.nodeName.toLowerCase() === 'li') {
                target.parentNode.classList.add('has-jax');
              }
            });
          }, '', false]
        }
      }
    };
    (function () {
      var script = document.createElement('script');
      script.src = '//cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js';
      script.defer = true;
      document.head.appendChild(script);
    })();
  } else {
    MathJax.startup.document.state(0);
    MathJax.texReset();
    MathJax.typeset();
  }
</script>

    

  

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : '02ff9c675de13b9fbbbb',
      clientSecret: 'f56e4a1eb98b4f3dfde62cca3ec7c19f9fd6a016',
      repo        : 'gitalk-comments',
      owner       : 'geng-lee',
      admin       : ['geng-lee'],
      id          : 'be5a3340e738836ee2a997b53f195483',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
